// @flywire/react-hooks/dist/umd/hooks/useCookie/cookies/cookies.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getCookie = exports.deleteCookie = exports.setCookie = exports.DELETE_COOKIE_EXPIRATION = exports.PERMANENT_COOKIE_EXPIRATION = void 0;
    exports.PERMANENT_COOKIE_EXPIRATION = '01 Aug 9999 00:00:00 UTC';
    exports.DELETE_COOKIE_EXPIRATION = '01 Aug 1970 00:00:00 UTC';
    function findCookie(name) {
        const result = document.cookie
            .split(';')
            .map((cookie) => cookie.trim())
            .filter((cookie) => cookie.startsWith(`${name}=`));
        return result.length ? result[0] : undefined;
    }
    function extractValue(cookie, name) {
        return decodeURIComponent(cookie
            .substring(name.length + 1)
            .split('+')
            .join(' '));
    }
    function setCookie(name, value, opts = {
        permanent: false,
    }) {
        const { permanent, domain } = opts;
        const cookieExpiration = permanent
            ? `expires=${exports.PERMANENT_COOKIE_EXPIRATION}`
            : null;
        const cookieDomain = domain ? `domain=${domain}` : null;
        const cookie = [`${name}=${value}`, cookieExpiration, 'path=/', cookieDomain]
            .filter(Boolean)
            .join('; ');
        document.cookie = cookie;
    }
    exports.setCookie = setCookie;
    function deleteCookie(name) {
        const expires = exports.DELETE_COOKIE_EXPIRATION;
        const cookie = `${name}=''; expires=${expires}; path=/`;
        document.cookie = cookie;
    }
    exports.deleteCookie = deleteCookie;
    function getCookie(name) {
        const cookie = findCookie(name);
        if (!cookie) {
            return null;
        }
        return extractValue(cookie, name);
    }
    exports.getCookie = getCookie;
});
