// @flywire/react-hooks/dist/esm/hooks/useCookie/cookies/cookies.js
export const PERMANENT_COOKIE_EXPIRATION = '01 Aug 9999 00:00:00 UTC';
export const DELETE_COOKIE_EXPIRATION = '01 Aug 1970 00:00:00 UTC';
function findCookie(name) {
    const result = document.cookie
        .split(';')
        .map((cookie) => cookie.trim())
        .filter((cookie) => cookie.startsWith(`${name}=`));
    return result.length ? result[0] : undefined;
}
function extractValue(cookie, name) {
    return decodeURIComponent(cookie
        .substring(name.length + 1)
        .split('+')
        .join(' '));
}
export function setCookie(name, value, opts = {
    permanent: false,
}) {
    const { permanent, domain } = opts;
    const cookieExpiration = permanent
        ? `expires=${PERMANENT_COOKIE_EXPIRATION}`
        : null;
    const cookieDomain = domain ? `domain=${domain}` : null;
    const cookie = [`${name}=${value}`, cookieExpiration, 'path=/', cookieDomain]
        .filter(Boolean)
        .join('; ');
    document.cookie = cookie;
}
export function deleteCookie(name) {
    const expires = DELETE_COOKIE_EXPIRATION;
    const cookie = `${name}=''; expires=${expires}; path=/`;
    document.cookie = cookie;
}
export function getCookie(name) {
    const cookie = findCookie(name);
    if (!cookie) {
        return null;
    }
    return extractValue(cookie, name);
}
