// @flywire/react-hooks/dist/cjs/hooks/useFormState/useFormState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
function useFormState(initialValues = {}) {
    const [state, setState] = (0, react_1.useState)({
        values: initialValues,
        dirtyFields: {},
    });
    function update(name, value) {
        setState(Object.assign(Object.assign({}, state), { values: Object.assign(Object.assign({}, state.values), { [name]: value }), dirtyFields: Object.assign(Object.assign({}, state.dirtyFields), { [name]: true }) }));
    }
    function updateAll(newFields = {}) {
        setState(Object.assign(Object.assign({}, state), { values: Object.assign(Object.assign({}, state.values), newFields), dirtyFields: Object.assign(Object.assign({}, state.dirtyFields), Object.keys(newFields).reduce((acc, item) => {
                acc[item] = true;
                return acc;
            }, {})) }));
    }
    function reset(values = initialValues) {
        setState({ values, dirtyFields: {} });
    }
    return Object.assign(Object.assign({}, state), { update, updateAll, reset });
}
exports.default = useFormState;
