// @flywire/react-hooks/dist/cjs/hooks/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useValidate = exports.useToggle = exports.useThrottle = exports.useStep = exports.useOnScroll = exports.useOnClickOutside = exports.useLocalStorageState = exports.useIdleTimer = exports.useFormState = exports.useDebounce = exports.useCookie = void 0;
var useCookie_1 = require("./useCookie");
Object.defineProperty(exports, "useCookie", { enumerable: true, get: function () { return __importDefault(useCookie_1).default; } });
var useDebounce_1 = require("./useDebounce");
Object.defineProperty(exports, "useDebounce", { enumerable: true, get: function () { return __importDefault(useDebounce_1).default; } });
var useFormState_1 = require("./useFormState");
Object.defineProperty(exports, "useFormState", { enumerable: true, get: function () { return __importDefault(useFormState_1).default; } });
var useIdleTimer_1 = require("./useIdleTimer");
Object.defineProperty(exports, "useIdleTimer", { enumerable: true, get: function () { return __importDefault(useIdleTimer_1).default; } });
var useLocalStorageState_1 = require("./useLocalStorageState");
Object.defineProperty(exports, "useLocalStorageState", { enumerable: true, get: function () { return __importDefault(useLocalStorageState_1).default; } });
var useOnClickOutside_1 = require("./useOnClickOutside");
Object.defineProperty(exports, "useOnClickOutside", { enumerable: true, get: function () { return __importDefault(useOnClickOutside_1).default; } });
var useOnScroll_1 = require("./useOnScroll");
Object.defineProperty(exports, "useOnScroll", { enumerable: true, get: function () { return __importDefault(useOnScroll_1).default; } });
var useStep_1 = require("./useStep");
Object.defineProperty(exports, "useStep", { enumerable: true, get: function () { return __importDefault(useStep_1).default; } });
var useThrottle_1 = require("./useThrottle");
Object.defineProperty(exports, "useThrottle", { enumerable: true, get: function () { return __importDefault(useThrottle_1).default; } });
var useToggle_1 = require("./useToggle");
Object.defineProperty(exports, "useToggle", { enumerable: true, get: function () { return __importDefault(useToggle_1).default; } });
var useValidate_1 = require("./useValidate");
Object.defineProperty(exports, "useValidate", { enumerable: true, get: function () { return __importDefault(useValidate_1).default; } });
